README
Getopt::Auto is Copyright (C) 2010, Geoffrey Leach
Getopt::Auto is Copyright (C) 2003-2009, Simon Cozens

Getopt::Auto version 1.9.4
========================

Getopt:Auto automates the processing of command-line options
for Perl scripts.

INSTALLATION

To install this module type the following:

   perl Build.PL
   Build
   Build test
   Build install

DEPENDENCIES

Please check Build.PL for this information.

LICENSE INFORMATION

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl 5.10.1.  For more details,
see the full text of the licenses at
    <http://www.perlfoundation.org/artistic_license_1_0>,
and
    <http://www.gnu.org/licenses/gpl-2.0.html>.





Changes
Revision history for Perl extension Getopt::Auto.

1.0   Fri Jan  3 12:33:09 2003
	- original version; created by h2xs 1.21 with options
		-XA -n Getopt::Auto

1.9.0 Thu Dec  3 16:42:08 PST 2009
    Completed according to original intent of incomplete
    version 1.0. Version is '1.9.0' in anticipation of going to
    2.0 when bugs are worked out.

1.9.1 Sun Jan 10 15:49:08 PST 2010
    Builtin help does not exit 0. Change 'exists' to 'defined' in test

1.9.1 Mon Jan 18 08:28:57 PST 2010
    Windows failure in test t/02-internals_magic.t
    Construct the path using catfile() to avoid the problem of '/' vs '\'

1.9.1 Mon Jan 18 08:31:32 PST 2010
    Failures of Spec build on 5.6.x and 5.8.1 Perls
    Mis-application of each() (see `perldoc -f each`). Result is random.
    Patch courtesy of Bruce Gray <bruce.gray@acm.org>.

1.9.2 Thu Jul  1 16:06:58 PDT 2010
    Remove _ from _test_option so that perlcritic does not complain.
    It is, after all, intended for external use.
    Fix bug where command line with _no_ options, but args caused
    objection. In process made the default option type BARE.

1.9.3 Tue Oct  5 11:27:02 PDT 2010
    Add automatic help for unregistered options, so that you can
    avoid having to run the program a second time with --help (or
    is that -h? Oh, bother!)

1.9.4 Wed Dec 22 15:25:28 PST 2010
    Added comments.
    Tue Jul 26 20:13:22 PDT 2011
    Repair for 69822 (=begin/=end stopwords need blank lines)

1.9.5 Tue Jul 26 20:55:31 PDT 2011
    Fixed un-commented DB::
    Added example for uage()


t/Internals_magic_multi.pm
#! /usr/bin/perl

#  Copyright (C) 2010, Geoffrey Leach
#
#===============================================================================
#
#         FILE:  02-internals_magic_multi.pm
#
#  DESCRIPTION:  Test the construction of internal data structures
#                which result from the "magic" mode of Getopt::Auto
#                when running multiple files
#
#       AUTHOR:  Geoffrey Leach (), <geoff@hughes.net>
#      VERSION:  1.9.4
#      CREATED:  07/06/2009 03:27:58 PM PDT
#===============================================================================

use strict;
use warnings;

use Test::Output;

package Internals_magic_multi;
use Getopt::Auto;

use 5.006;
our $VERSION = '1.9.4';

## no critic (RequireCheckedSyscalls)

sub internals_magic_multi_pm {
    print "did Internals_magic_multi_pm\n";
    return;
}

1;


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
	module_name        => 'Getopt::Auto',
	license            => 'perl',
	dist_author        => 'Geoffrey Leach <geoff@hughes.net>',
	requires           => {},
	build_requires     => { 'Test::More' => 0,
                            'Test::Output' => 0,
                            'File::Spec' => 0,
                            'File::Basename' => 0,
                            'File::Spec::Functions' => 0,
                            'Readonly' => 0,
                            'Pod::Usage' => 0,
                          },
	add_to_cleanup     => [ 'Getopt-Auto-*' ],
 	create_makefile_pl => 'traditional',
);

$builder->create_build_script();
#$builder->dispatch('build');
#$builder->dispatch('test', verbose => 1);
#$builder->dispatch('install')


scripts/tour.pl
#!/usr/bin/perl 
#===============================================================================
#
#         FILE:  tour.pl
#
#        USAGE:  ./tour.pl
#
#  DESCRIPTION:  A Tour of Getopt::Auto's facilities
#
#       AUTHOR:  Geoffrey Leach (), geoff@hughes.net
#      VERSION:  1.0
#      CREATED:  09/06/2009 04:37:59 PM
#===============================================================================

use strict;
use warnings;

# OK, so I lied. We're only going to tour the "magic" of Getopt::Auto.
# If you need to use the list parameters, you should be able to
# backfit them from what follows.  Or not.

use Getopt::Auto( { 'nobare' => 1, 'init' => \&init, 'okerror' => 1 } );
use Pod::Usage;
use 5.006;

# Checking the return vaue from a print statement?
# Give me a break!
## no critic (RequireCheckedSyscalls)

# Here's the version that Getopt::Auto will look for.
# "our" is not necessary, as the value is obtained by scanning the source.
# However, perlcritic complains if it's _not_ there. So ...
our $VERSION = "1.0";

# This is where the single-use options show up.
# Notice the "our". It's necessary. Perlcritic complains! Sigh!

our %options;    ## no critic (ProhibitPackageVars)

# These subs are here only because that's how I like to organize.
# However, you could say that they are here because they are
# executed by Getopt::Auto (if called out by options) before the
# actual execution starts. Now, how cool is that?

my $did_itemopt;

sub init {
    print "Did init\n";
    return;
}

sub itemopt {
    my $itemval = shift @ARGV;
    if ( $itemval =~ m{^\d+$/}sxm ) {
        print "Your itemopt value is: $itemval\n";
    }
    else {
        print "Itemopt value $itemval is not integer\n";
    }
    $did_itemopt = 1;
    delete $options{'--itemopt'};
    return;
}

sub usage {
    pod2usage( -verbose => 99, -sections => 'DESCRIPTION' );
    exit 0;
}

sub help {
    pod2usage( -verbose => 2 );
    exit 0;
}

# Program execution starts here, even though some of the
# subs may have been executed aready

# Process the single-use options
# The opions are deleted so that the "Undefined option"
# processing does not complain about them.

if ( ( !@ARGV ) && ( not keys %options ) && ( not defined $did_itemopt ) ) {
    print "What, no options?\n";
    exit 0;
}

if ( exists $options{'--nodef'} ) {
    print "You selected '--nodef', but it was not defined!\n";
    delete $options{'--nodef'};
}

if ( exists $options{'--cutup'} ) {
    print "You selected '--cutup'\n";
    delete $options{'--cutup'};
}

if ( exists $options{'-c'} ) {
    print "You selected '-c'\n";
    delete $options{'-c'};
}

foreach ( keys %options ) {
    print "Option: $_ => $options{$_}\n";
}

foreach (@ARGV) {
    print "Unused command-line data: $_\n";
}

exit 0;

__END__

=pod

=head1 SYNOPSIS

Shows the use of C<Getopt::Auto>.

    ./tour.pl <options>

=head1 DESCRIPTION

Some things to try.

The options listed below. Pretty obvious.

C<tour.pl> without any options. Notice how we check for that.
Notice C<use Getopt::Auto> has a hashref as argument. The C<okerror> allows
some additional processing after command-line errors (such as "not a registered option")
are detected.

C<tour.pl -food>: Notice how it's split up.

C<tour.pl --notexpected notfound - notwanted>. Notice that --notexpected (double dash) is left on the
command line, whereas -notexpected (single dash) would not be. "notfound" is something that might have been 
processed by --notexpected, had it been expected. "notwanted" would not be processed in
any case because it follows the "Cease and Desist" marker, "-".

C<tour.pl --itemopt 12 --itemopt xy> demonstrates multiple uses of a single option (not very
exciting) and how an option can process its values. Complexity limited only by your
imagination (and tolerance for user errors!)

C<tour.pl --item work>, just to prove that the documentation does not lie.

C<tour.pl --nodef>. What a surprise! Where's the =head2? 

=head1 OPTIONS

Ok, here we go.

=head2 -u, --usage - Prints the DESCRIPTION.

=head2 -h, --help - Overrides the builtin.

=head2 -c, --cutup - The c option

=head2 -f - The f option, part of the -food demo

Cutup (and c) do not have subs, so to see that they are called, we must
check %options.

As we're going to all this trouble to create POD, it seems a shame not
to use it. Note that help() is not called in the code. It happens before
anything else.

=head2 work - do the work!

This would be a bare option, except that we've said 'nobare'
in the configuration
to outlaw bareness. This is not a nude beach, after all.

=head2 --item options

Note that there is no ' - ', so not an option
What follows is an option that is introduced by "=item".

=over 4

=item --itemopt - Which is an option in an item

Note that there is no '-i'
This option has an integer argument.

=back

=begin stopwords
notfound
notwanted
builtin
nobare
itemopt
=end stopwords

=cut
